framework/src/times/Times.C

```diff
@@ -0,0 +1,124 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "Times.h"
+#include "libmesh/parallel_algebra.h"
+
+InputParameters
+Times::validParams()
+{
+  InputParameters params = GeneralReporter::validParams();
+  params.addParam<bool>(
+      "unique_times", true, "Whether duplicate values should be removed from the Times vector");
+  params.addParam<Real>(
+      "unique_tolerance",
+      1e-12,
+      "Absolute tolerance for removing duplicated times when using setting unique_times to true");
+  params.addParam<bool>("auto_sort", true, "Whether Times should be sorted");
+  // This parameter should be set by each derived class depending on whether the generation of
+  // times is replicated or distributed. We want times to be replicated across all ranks
+  params.addRequiredParam<bool>("auto_broadcast",
+                                "Wether Times should be broadcasted across all ranks");
+  params.addParamNamesToGroup("auto_broadcast auto_sort unique_times unique_tolerance", "Advanced");
+  params.registerBase("Times");
+  return params;
+}
+
+Times::Times(const InputParameters & parameters)
+  : GeneralReporter(parameters),
+    // Times will be replicated on every rank
+    _times(declareValueByName<std::vector<Real>, ReporterVectorContext<Real>>(
+        "times", REPORTER_MODE_REPLICATED)),
+    _need_broadcast(getParam<bool>("auto_broadcast")),
+    _need_sort(getParam<bool>("auto_sort")),
+    _need_unique(getParam<bool>("unique_times")),
+    _unique_tol(getParam<Real>("unique_tolerance"))
+{
+}
+
+Real
+Times::getTimeAtIndex(unsigned int index) const
+{
+  if (_times.size() < index)
+    mooseError("Times retrieved with an out-of-bound index");
+
+  if (_times.size())
+    return _times[index];
+  else
+    mooseError("Times vector has not been initialized.");
+}
+
+Real
+Times::getPreviousTime(const Real current_time) const
```

Comment:
Given that `_times` is a sequence, there could in theory be multiple "previous" times for a given time, right? I would put this somewhere in the documentation